File: frontend/src/features/ore-veins/api/oreVeins.ts
import { useQuery } from '@tanstack/react-query'

import { API_BASE } from '../../../shared/api'

/** A GregTech ore vein cached by Visual Prospecting. */
export interface OreVein {
  x: number
  z: number
  cx: number
  cz: number
  kind: string // VP palette name, e.g. "ore.mix.gold"
  depleted: boolean
  // Enriched server-side from the bundled ore-vein registry; absent for an
  // unknown/modded vein or when no dump is available (frontend falls back).
  name?: string | null // GTNH display name, e.g. "Magnetite & Gold"
  rgb?: number | null // representative material colour, 0xRRGGBB
  texture?: string | null // ore-texture key (best-effort; usually null)
}

export interface OreVeinsResponse {
  available: boolean // false = this world has no Visual Prospecting data
  veins: OreVein[]
  // Deduped ore-overlay PNGs (base64), keyed by icon name — a vein's `texture` is
  // a key here. Tint by the vein `rgb` to render. Absent/empty on an old dump
  // (frontend falls back to a flat colour dot).
  sprites?: Record<string, string>
  // Sprite keys that are pre-coloured (drawn as-is, no rgb tint) — e.g. gold/iron,
  // whose art lives in the overlay layer rather than a tintable grayscale base.
  sprites_precolored?: string[]
}

async function fetchOreVeins(dimensionPath: string): Promise<OreVeinsResponse> {
  const res = await fetch(
    `${API_BASE}/worlds/ore-veins?world_path=${encodeURIComponent(dimensionPath)}`
  )
  if (!res.ok) throw new Error(`Failed to load ore veins (${res.status})`)
  return (await res.json()) as OreVeinsResponse
}

/**
 * Ore veins Visual Prospecting has cached for this dimension. Enabled lazily so
 * the (cheap) read only runs once the overlay is turned on for a dimension.
 */
export function useOreVeins(dimensionPath: string | null, enabled: boolean) {
  return useQuery({
    queryKey: ['oreVeins', dimensionPath],
    queryFn: () => fetchOreVeins(dimensionPath!),
    enabled: !!dimensionPath && enabled,
  })
}

File: backend/app/models/ore_veins.py
from pydantic import BaseModel


class OreVein(BaseModel):
    """A GregTech ore vein cached by Visual Prospecting."""

    x: int  # world X of the vein anchor (jump/marker target)
    z: int  # world Z
    cx: int  # vein chunk coords (as stored by VP)
    cz: int
    kind: str  # VP palette name, e.g. "ore.mix.gold"
    depleted: bool  # true once the vein has been mined out
    # Enriched from the bundled ore-vein registry (ore_vein_dump.json). None when
    # this kind isn't in the registry (an unknown/modded vein, or no dump present) —
    # the frontend then falls back to its built-in name/colour table.
    name: str | None = None  # GTNH display name, e.g. "Magnetite & Gold"
    rgb: int | None = None  # representative material colour as 0xRRGGBB
    texture: str | None = None  # ore-texture key (best-effort; usually None)


class OreVeinsResponse(BaseModel):
    # False when this world has no Visual Prospecting data (the mod wasn't used,
    # or nothing was cached for this dimension) — the UI shows a hint instead.
    available: bool
    veins: list[OreVein]
    # Deduped ore-overlay PNGs (base64), keyed by icon name; each vein's ``texture``
    # is a key here. The frontend tints a sprite by the vein ``rgb`` (as GregTech
    # does at render). Empty when no dump/sprites are available, in which case the
    # frontend falls back to a flat colour dot.
    sprites: dict[str, str] = {}
    # Sprite keys that are pre-coloured (a material's ore art lives in its overlay,
    # not a tintable grayscale base — e.g. gold/iron/copper). The frontend draws
    # these as-is (no rgb tint), matching how GregTech renders them.
    sprites_precolored: list[str] = []
